blogs/docker-ci-tricks-2/blog.md

```diff
@@ -0,0 +1,209 @@
+
+# Nifty Docker tricks for your CI (vol. 2)
+
+The [first part of the series](https://medium.com/virtuslab/nifty-docker-tricks-for-your-ci-vol-1-c4a36d2192ea) outlined the techniques for reducing the image size,
+as well as avoiding the need to rebuild the image every time when the codebase changes.
+Let's continue with further Docker-related CI tricks, as showcased in [git-machete](https://github.com/VirtusLab/git-machete)'s CircleCI setup.
+
+## Caching the images: make use of Docker tags
+
+It would be nice to cache the generated images, so that CI doesn't build the same stuff over and over again.
+By the way, the purpose of the extreme image size optimization outlined [in the previous part](https://medium.com/virtuslab/nifty-docker-tricks-for-your-ci-vol-1-c4a36d2192ea) is also to facilitate caching.
+
+Think for a moment what exactly causes one generated image to be different from another.
+We obviously have to take into account the versions of Git and Python &mdash; passing different combinations of these will surely result in a different final image.
+Therefore, the Docker image tag should include the information about these versions in order to make the caching possible.
+
+But with respect to the project files... since it's `ci/tox/build-context` that's passed as build context,
+the Dockerfile doesn't know anything about the files from outside that directory!
+This means that even though other files change as the project is being developed,
+we could theoretically use the same image if `ci/tox/build-context` remains untouched.
+Note that the build context usually changes very rarely compared to the actual source files.
+
+There's a catch here, though.
+The build context is not the only thing that can affect the final Docker image.
+A change to the `Dockerfile` itself can obviously lead to a different result,
+but modification to `docker-compose.yml` and even the scripts that run `docker-compose` may also influence the result e.g. by changing environment variables and/or values passed as arguments.
+But since all those files reside under `ci/tox/`, this doesn't make things much worse &mdash;
+the entire resulting build image depends only on the contents of `ci/tox/` directory instead of `ci/tox/build-context/`
+(with one exception that packages installed via `apt-get` can also get updated over time in their respective APT repositories).
```

Comment:
Whoops an old error of mine, not spotted by anyone despite [you-can-see-yourself-how-many rounds of review](https://github.com/PawelLipski/docker-ci-blog/commits/master/blog.md) 😅

Reply:
It's not `apt` since it's Alpine Linux